### PR TITLE
Made set_attenuation_coefficients private

### PR DESCRIPTION
**Pull request by Abinash-bit**
### Summary

This pull request resolves conflicts and merges the latest changes from the main branch into the feature-branch. The changes include:

- Integrated the logic of `set_attenuation_coefficients` directly into the `fnrgf` function.
- Updated the `fnrgf` function to accept `range_mean`, `range_std`, and `cutoff` parameters.
- Resolved conflicts in `radial.py` and `test_radial.py`.

### Changes Made

- Merged the latest changes from the main branch.
- Resolved conflicts in the `fnrgf` function.
- Updated the test cases in `test_radial.py` to reflect the changes.

### Testing

- All tests in `test_radial.py` have been updated and pass successfully.

### Related Issues

- Fixes #239  

### Checklist

- [x] Code follows the project's coding standards.
- [x] Documentation has been updated.
- [x] Tests have been added or updated to cover the changes.
- [x] All tests pass successfully.

### Reviewers

- @nabobalis

### CODE REVIEW

**Comment by Abinash-bit** [comment]
Hey @nabobalis please have a look.

**Comment by nabobalis** [comment]
So this replaces the original pull request?

File: sunkit_image/radial.py [hunk 1]
```diff
@@ -466,10 +407,14 @@ def fnrgf(
     order : `int`, optional
         Order (number) of fourier coefficients and it can not be lower than 1.
         Defaults to 3.
-    attenuation_coefficients : `float`
-        A two dimensional array of shape ``[2, order + 1]``. The first row contain attenuation
-        coefficients for mean calculations. The second row contains attenuation coefficients
-        for standard deviation calculation.
+    range_mean : `list`, optional
+        A list of length of ``2`` which contains the highest and lowest values between which the coefficients for
+        mean approximation be calculated in a linearly decreasing manner.
```
Comment (nabobalis) [inline sunkit_image/radial.py:412]: The default values are missing for each of the new parameters.

File: sunkit_image/radial.py [hunk 2]
```diff
@@ -466,10 +407,14 @@ def fnrgf(
...
+    range_std : `list`, optional
```
Comment (nabobalis) [inline sunkit_image/radial.py:413]: The names should be changed as they were in the original pull request.

File: sunkit_image/radial.py [hunk 3]
```diff
@@ -360,75 +362,14 @@ def nrgf(
     return new_map
 
 
-def set_attenuation_coefficients(order, range_mean=None, range_std=None, cutoff=0):
```
Comment (nabobalis) [inline sunkit_image/radial.py:363]: Can we just keep this private and have the function call to this?

I don't see the upside to adding this code directly to the main function.

File: sunkit_image/tests/test_radial.py [hunk 4]
```diff
@@ -233,30 +239,6 @@ def test_multifig_rhef(aia_171_map):
     return fig
 
 
-def test_set_attenuation_coefficients():
```
Comment (nabobalis) [inline sunkit_image/tests/test_radial.py:236]: If we restore the original function, I want it tested still.

File: sunkit_image/radial.py [hunk 5]
```diff
@@ -130,6 +130,69 @@ def _percentile_ranks_numpy_inplace(arr):
         raise NotImplementedError(msg)
     return ranking_func
 
+def _set_attenuation_coefficients(order, range_mean=None, range_std=None, cutoff=0):
```
Comment (nabobalis) [inline sunkit_image/radial.py:133]: Can this be moved back to its original location?

Comment (Abinash-bit) [reply sunkit_image/radial.py:133]: ok.

File: sunkit_image/radial.py [hunk 6]
```diff
@@ -130,6 +130,9 @@ def _percentile_ranks_numpy_inplace(arr):
         raise NotImplementedError(msg)
     return ranking_func
 
+
+
+
```
Comment (nabobalis) [inline sunkit_image/radial.py:135]:
```suggestion


```

File: sunkit_image/tests/test_radial.py (hunk 4 above)
Comment (nabobalis) [inline sunkit_image/tests/test_radial.py:236]: This needs to be resorted and updated to use the renamed function.

File: sunkit_image/radial.py [hunk 7]
```diff
@@ -466,10 +471,17 @@ def fnrgf(
     order : `int`, optional
         Order (number) of fourier coefficients and it can not be lower than 1.
         Defaults to 3.
-    attenuation_coefficients : `float`
-        A two dimensional array of shape ``[2, order + 1]``. The first row contain attenuation
-        coefficients for mean calculations. The second row contains attenuation coefficients
-        for standard deviation calculation.
+    range_mean : `list`, optional
```
Comment (nabobalis) [inline sunkit_image/radial.py:474]: These were renamed in the original PR and that should happen here too.

**Comment by nabobalis** [comment]
This will need a breaking changelog entry as well to round off the PR.

File: sunkit_image/radial.py (hunk 7 above)
Comment (Abinash-bit) [reply sunkit_image/radial.py:472]: do you suggest range_mean should be renamed to mean_attenuation_range?
and range_std to std_attenuation_range.

Comment (nabobalis) [reply sunkit_image/radial.py:472]: Yeah, those.

File: sunkit_image/tests/test_radial.py [hunk 8]
```diff
@@ -327,3 +309,27 @@ def test_intensity_enhance_errors(map_test1):
     scale = 1 * map_test1.rsun_obs
     with pytest.raises(ValueError, match="The fit range must be strictly increasing."):
         rad.intensity_enhance(map_test1, scale=scale, fit_range=fit_range[::-1])
+
+
+def test_set_attenuation_coefficients():
```
Comment (nabobalis) [inline sunkit_image/tests/test_radial.py:314]: Can this not be moved back to its original location?

**Comment by Abinash-bit** [comment]
> This will need a breaking changelog entry as well to round off the PR.

I see in the changelog folder there are .rst files  for example 185.breaking.rst what does 185 refers , like while making .rst file what should i take into consideration.

File: changelog/243.breaking.rst [hunk 9]
```diff
@@ -0,0 +1,29 @@
+The function ``set_attenuation_coefficients`` was renamed to ``_set_attenuation_coefficients`` to indicate that it is a helper function and not intended to be part of the public API in ``radial.py``.
```
Comment (nabobalis) [inline changelog/243.breaking.rst:1]: The goal of the changelog is to tell users what has been broken and how they can fix their code.

File: sunkit_image/radial.py [hunk 10]
```diff
@@ -426,9 +427,11 @@ def set_attenuation_coefficients(order, range_mean=None, range_std=None, cutoff=
 def fnrgf(
     smap,
     *,
-    attenuation_coefficients,
     radial_bin_edges=None,
     order=3,
+    mean_attenuation_range=None,
+    std_attenuation_range=None,
+    cutoff=0,
```
Comment (nabobalis) [inline sunkit_image/radial.py:434]:
```suggestion
    mean_attenuation_range=None,
    std_attenuation_range=None,
    cutoff=0,
    order=3,
```

File: sunkit_image/tests/test_radial.py [hunk 11]
```diff
@@ -327,3 +332,5 @@ def test_intensity_enhance_errors(map_test1):
     scale = 1 * map_test1.rsun_obs
     with pytest.raises(ValueError, match="The fit range must be strictly increasing."):
         rad.intensity_enhance(map_test1, scale=scale, fit_range=fit_range[::-1])
+
+
```
Comment (nabobalis) [inline sunkit_image/tests/test_radial.py:336]:
```suggestion
```

**Review by nabobalis — APPROVED** [review]
Just needs the changelog tweaking but it LGTM

**Comment by nabobalis** [comment]
```
Sphinx-Gallery successfully executed 10 out of 11 files subselected by:
WARNING: Here is a summary of the problems encountered when running the examples:

Unexpected failing examples (1):

    ../examples/radial_gradient_filters.py failed leaving traceback:

    Traceback (most recent call last):
      File "/home/runner/work/sunkit-image/sunkit-image/examples/radial_gradient_filters.py", line 51, in <module>
        attenuation_coefficients = radial.set_attenuation_coefficients(order)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    AttributeError: module 'sunkit_image.radial' has no attribute 'set_attenuation_coefficients'. Did you mean: '_set_attenuation_coefficients'?
```

Looks like one of the examples needs to be updated.

File: examples/radial_gradient_filters.py [hunk 12]
```diff
@@ -48,7 +48,7 @@
 # choose them according to your requirements.
 
 order = 20
-attenuation_coefficients = radial.set_attenuation_coefficients(order)
+attenuation_coefficients = radial._set_attenuation_coefficients(order)
```
Comment (nabobalis) [inline examples/radial_gradient_filters.py:51]: We do not want to encourage use of the private function, the rest of the example needs to be updated to use the new keywords.

**Comment by nabobalis** [comment]
Now just the changelog and the precommit left.

**Comment by Abinash-bit** [comment]
Screenshot 2024-11-08 112503

i can not get the error , like where is the actual error.

**Comment by nabobalis** [comment]
I think it would be easier to run the pre-commit locally and have it fix the issues for you instead of doing it manually.

**Comment by Abinash-bit** [comment]
Hey @nabobalis i think all the checks have passed please review.

File: examples/radial_gradient_filters.py [hunk 13]
```diff
@@ -41,16 +41,26 @@
...
 # choose them according to your requirements.
 
 order = 20
-attenuation_coefficients = radial.set_attenuation_coefficients(order)
-
-base_fnrgf = radial.fnrgf(aia_map, radial_bin_edges=radial_bin_edges, order=order, attenuation_coefficients=attenuation_coefficients, application_radius=1 * u.R_sun)
+attenuation_coefficients = [1.0 - i / order for i in range(order)]
+
+# Generate attenuation coefficients using _set_attenuation_coefficients
+mean_attenuation_range, std_attenuation_range = radial._set_attenuation_coefficients(order, attenuation_coefficients)
```
Comment (nabobalis) [inline examples/radial_gradient_filters.py:54]: Again, if showcase users using this function, there was no point in making it private. 

Either this is deleted, or we just make the function public but moved to utils.

**Comment by Abinash-bit** [comment]
Hey is this the correct way , apart from moving to utils ??

**Comment by Abinash-bit** [comment]
Hey @nabobalis i just pressed commit suggestions , now in the latest commit some test got failed , please help me to uncommit the latest commit.

File: changelog/243.breaking.rst [hunk 14]
```diff
@@ -0,0 +1,3 @@
+The function ``set_attenuation_coefficients`` has been removed from the public API.
+
+Now you can pass in the inputs used by ``set_attenuation_coefficients`` directly into `sunkit.radia.fnrgf`, which now accepts ``mean_attenuation_range`` and ``std_attenuation_range``.
```
Comment (nabobalis) [inline changelog/243.breaking.rst:3]:
```suggestion
Now you can pass in the inputs used by ``set_attenuation_coefficients`` directly into `sunkit.radial.fnrgf`, which now accepts ``mean_attenuation_range`` and ``std_attenuation_range``.
```
I introduced a bad typo which broke it.

Comment (Abinash-bit) [reply changelog/243.breaking.rst:3]: oh.

File: examples/radial_gradient_filters.py [hunk 15]
```diff
@@ -41,16 +41,27 @@
...
+std_attenuation_range = [1.0, 0.0]
+
+base_fnrgf = radial.fnrgf(
+    aia_map,
+    radial_bin_edges=radial_bin_edges,
+    order=order,
+    mean_attenuation_range=mean_attenuation_range,
+    std_attenuation_range=std_attenuation_range,
+    application_radius=1 * u.R_sun
+)
```
Comment (nabobalis) [inline examples/radial_gradient_filters.py:64]:
```suggestion
base_fnrgf = radial.fnrgf(
    aia_map,
    radial_bin_edges=radial_bin_edges,
    order=order,
    application_radius=1 * u.R_sun
)
```

**Comment by Abinash-bit** [comment]
> Hey is this the correct way , apart from moving to utils ??

??

File: examples/radial_gradient_filters.py [hunk 16]
```diff
@@ -41,16 +41,27 @@
 base_nrgf = radial.nrgf(aia_map, radial_bin_edges=radial_bin_edges, application_radius=1 * u.R_sun)
 
 ###########################################################################
-# We will need to work out  a few parameters for the FNRGF.
+# We will need to work out a few parameters for the FNRGF.
 #
 # Order is the number of Fourier coefficients to be used in the approximation.
-# The attenuation coefficient are calculated to be linearly decreasing, you should
+# The attenuation coefficients are calculated to be linearly decreasing, you should
 # choose them according to your requirements.
```
Comment (nabobalis) [inline examples/radial_gradient_filters.py:48]:
```suggestion
# The attenuation coefficients are calculated to be linearly decreasing, you should
# choose them according to your requirements. These can be changed by tweaking the following keywords: ``mean_attenuation_range`` and ``std_attenuation_range``.
```

**Comment by nabobalis** [comment]
> Hey @nabobalis i just pressed commit suggestions , now in the latest commit some test got failed , please help me to uncommit the latest commit.

We don't need to uncommit, we just need to fix it in another commit.

File: changelog/243.breaking.rst [hunk 17]
```diff
@@ -0,0 +1,3 @@
+The function ``set_attenuation_coefficients`` has been removed from the public API.
+
+Now you can pass in the inputs used by ``set_attenuation_coefficients`` directly into `sunkit.radial.fnrgf`, which now accepts ``mean_attenuation_range`` and ``std_attenuation_range``.
```
Comment (nabobalis) [inline changelog/243.breaking.rst:3]:
```suggestion
Now you can pass in the inputs used by ``set_attenuation_coefficients`` directly into `sunkit_image.radial.fnrgf`, which now accepts ``mean_attenuation_range`` and ``std_attenuation_range``.
```

**Comment by nabobalis** [comment]
Thank you very much @Abinash-bit

**Comment by Abinash-bit** [comment]
Thank you so much for your guidance in every steps @nabobalis .

**Pull request merged**